STREAMLINE-381 Test Source, Test Sink components

HeartSaVioR opened the pull request:
This is the implementation of design doc. for STREAMLINE-381.

More details here: [STREAMLINE-381](https://hwxiot.atlassian.net/browse/STREAMLINE-381) 

Design doc is here:
https://docs.google.com/a/hortonworks.com/document/d/1wt1OqhE6dQ9C_pzgXxRrn0lWCAhiMj8WBXlKgSSpYTQ/edit?usp=sharing

I'm copying API endpoint section to let UI team refers:

API Endpoints
----

> POST /api/v1/catalog/topologies/:topologyId/actions/testrun

>> Description

The endpoint runs given topology on test mode, and return the result of test run. Implementation details for running topology on test mode will be covered later.

>> Input parameters

* topologyId (Long)
* JSON String payload containing either test case ID or test records, and expected output records (optional)

```
{
  “testCaseId”: 1,
  “testRecords”: {
    “sourceA”: {
      “outputStreamA”: [
        {“driverId”: 24, “truckId”: 97, “eventTime”: “00:13.0”, ...},
        {“driverId”: 10, “truckId”: 85, “eventTime”: “00:12.2”, ...},
        …
      ],
      ...
    },
    “sourceB”: {
       …
    }
  },
  “expectedOutputRecords”: {
    "SINK1": [
      {"driverId": 20, "eventCount": 3, "maxLatitude": 39.78, ... },
      …
    ],
    ...
  },
}
```

Either “testCaseId” or “testRecords” should be provided. If “testCaseId” is provided, we pick specific test case instead of getting “testRecords” field, and don’t store new test case. If not, we store the test records to the test case.
"expectedOutputRecords" is optional. Only need to be provided when users want to compare actual output with expected output.

Assertion: All data sources should be covered. It will return HTTP status 400 if the assertion fails.

>> Output format (success case)

On succeed case, the response shows history entity of current test run.

```
{
  "id": 1,
  "topologyId": 1,
  "testCaseId": 1,
  "finish": true,
  "expectedOutputRecords": {
    "SINK1": [
      {"driverId": 20, "eventCount": 3, "maxLatitude": 39.78, ... },
      …
    ],
    ...
  },
  "actualOutputRecords": {
    "SINK1": [
      {"driverId": 20, "eventCount": 3, "maxLatitude": 39.78, ... },
      …
    ],
    ...
  },
  "startTime": 1234567890,
  "finishTime": 2345678901,
  "timestamp": 2345678901
}
```

> GET /api/v1/catalog/topologies/:topologyId/actions/testrun/histories

>> Description

The endpoint retrieves the histories of topology test run for given topology, and return recent N history entities.

>> Input parameters

* topologyId (Long)
* limit (Int) - optional, will fix the default value

>> Output format (success case)

On succeed case, the response shows history entities matched to given criteria.

```
[
  {
    "id": 99,
    "topologyId": 1,
    "testCaseId": 1,
    "finish": false,
    "expectedOutputRecords": {
      "SINK1": [
        {"driverId": 20, "eventCount": 3, "maxLatitude": 39.78, ... },
        …
      ],
      ...
    },
    "actualOutputRecords": {
      "SINK1": [
        {"driverId": 20, "eventCount": 3, "maxLatitude": 39.78, ... },
        …
      ],
      ...
    },
    “matched” : true,
    "startTime": 1234567890,
    "finishTime": 2345678901,
    "timestamp": 2345678901
  },
  …
]
```

> GET /api/v1/catalog/topologies/:topologyId/actions/testrun/testcases

>> Description

The endpoint retrieves the test cases user passed for previous test runs for given topology, and return recent N test cases.

>> Input parameters

* topologyId (Long)
* limit (int) - optional, will fix the default value

>> Output format (success case)

```
[
  {
    “Id”: 99,
    “topologyId”: 1,
    “records”: {
      “sourceA”: {
        “outputStreamA”: [
          {“driverId”: 24, “truckId”: 97, “eventTime”: “00:13.0”, ...},
          {“driverId”: 10, “truckId”: 85, “eventTime”: “00:12.2”, ...},
          …
        ],
        ...
      },
      “sourceB”: {
        …
      }
    },
    “timestamp”: 1234567890
  },
  …
]
```

The design doc. also describes table schema. If someone can't open design doc but needs table schema to work with, please let me know.

harshach commented:
@shahsank3t  please take a look at the APIs and let us know if you need any changes.

shahsank3t commented:
@HeartSaVioR 
It seems like due to some recent changes in Registry & Streamline, we are running into compilation failure for this PR. Can you please check once. 
Following are the logs:
Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.2:compile (default-compile) on project streamline-catalog: Compilation failure: Compilation failure:
[ERROR] /home/sameer/workspace/hortonworks/streams/streams/catalog/src/main/java/com/hortonworks/streamline/streams/catalog/TopologyTestRunHistory.java:[202,19] getSchema() in com.hortonworks.streamline.streams.catalog.TopologyTestRunHistory cannot implement getSchema() in com.hortonworks.streamline.storage.Storable
[ERROR] return type com.hortonworks.streamline.common.Schema is not compatible with com.hortonworks.registries.common.Schema
[ERROR] /home/sameer/workspace/hortonworks/streams/streams/catalog/src/main/java/com/hortonworks/streamline/streams/catalog/TopologyTestRunHistory.java:[56,31] incompatible types: java.util.Map<com.hortonworks.streamline.common.Schema.Field,java.lang.Object> cannot be converted to java.util.Map<com.hortonworks.registries.common.Schema.Field,java.lang.Object>
[ERROR] /home/sameer/workspace/hortonworks/streams/streams/catalog/src/main/java/com/hortonworks/streamline/streams/catalog/TopologyTestRunHistory.java:[201,5] method does not override or implement a method from a supertype
[ERROR] /home/sameer/workspace/hortonworks/streams/streams/catalog/src/main/java/com/hortonworks/streamline/streams/catalog/TopologyTestRunCase.java:[51,31] incompatible types: java.util.Map<com.hortonworks.streamline.common.Schema.Field,java.lang.Object> cannot be converted to java.util.Map<com.hortonworks.registries.common.Schema.Field,java.lang.Object>

Thanks in advance.

HeartSaVioR commented:
@shahsank3t Yes I'll fix it and mention you soon.

HeartSaVioR commented:
@shahsank3t Just rebased.

priyank5485 commented on bootstrap/sql/mysql/create_tables.sql:
```diff
@@ -352,3 +352,31 @@ CREATE TABLE IF NOT EXISTS component (
+
+CREATE TABLE IF NOT EXISTS topology_test_run_histories (
+  id BIGINT AUTO_INCREMENT NOT NULL,
+  topologyId BIGINT NOT NULL,
```
If we have a fk reference to topology_test_run_case id then we can get rid of topologyId here since an entry in that table will provide topologyId and version. If its kept here only for convenience i am fine with it

HeartSaVioR replied:
It's for convenience, and actually we can't rely on joining even if we define FK, given that we don't abstract join operation between two tables. So for Streamline (and Registry), FK is only for checking integrity, nothing more.

HeartSaVioR replied:
Maybe we could (and might be better to) replace our own implementation of ORM with popular and well defined thing like Hibernate. I'm not sure it can cover both RDB and Phoenix though.

priyank5485 commented on bootstrap/sql/mysql/create_tables.sql:
```diff
@@ -352,3 +352,31 @@ CREATE TABLE IF NOT EXISTS component (
+  success CHAR(5) NOT NULL,
+  expectedOutputRecords TEXT,
+  actualOutputRecords TEXT,
+  matched CHAR(5),
```
Do we need both success and matched columns? I did a quick search and saw setMatched being called. But setSuccess was not called anywhere.

HeartSaVioR replied:
It's being set from `finishSuccessfully` and `finishWithFailures`. `setSuccess` is available only for getter/setter pair.

priyank5485 replied:
Missed that part. Looks good

priyank5485 commented on streams/catalog/src/main/java/com/hortonworks/streamline/streams/catalog/service/StreamCatalogService.java:
```diff
@@ -2341,4 +2343,55 @@ private String getTopologyComponentBundleJarName (TopologyComponentBundle topolo
         return dao.find(File.NAME_SPACE, queryParams);
     }
+
+    public Collection<TopologyTestRunHistory> listTopologyTestRunHistory(Long topologyId) {
```
Does it make sense to add version id as well to some of these methods that have topology id as argument?

HeartSaVioR replied:
We need to review previous test history for previous version as well to compare why test fails on current version. 
And if my understanding is right, same topology version should have same topology DAG so test result should be same except end-user changes expected output, which gives no benefit.

priyank5485 replied:
May be i did not put it clearly. I meant should we be providing an api to return results for a particular version of a topology? Presumption is they could be running multiple tests with same version of topology. Let me know if i am misunderstanding something.

HeartSaVioR replied:
Ah OK you meant additional API. That makes sense. Will address.

priyank5485 commented on streams/runners/storm/actions/src/main/java/com/hortonworks/streamline/streams/actions/storm/topology/TestTopologyDagCreatingVisitor.java:
```diff
@@ -0,0 +1,147 @@
+
+    @Override
+    public void visit(Edge edge) {
+        // no-op
```
Instead of having copyEdges called from visit (StreamlineProcessor/StreamlineSink) we can probably just do it in visit(Edge). And anywhere from/to is a source or a sink we replace it with test source and sink. It might not be possible or i might be missing something. But if you think its better and easy to change we should change it

HeartSaVioR replied:
I gave a first try with that, but realized it's not easy.
To copy Edge we need to know about source and destination, and if either (or both) source or destination is source or sink, it should connect with replaced thing, which is not easy as it seems.

priyank5485 replied:
Ok

priyank5485 commented on streams/runners/storm/layout/src/main/java/com/hortonworks/streamline/streams/layout/storm/TestRunSinkBoltFluxComponent.java:
```diff
@@ -0,0 +1,51 @@
+    @Override
+    protected void generateComponent() {
+        testRunSink = (TestRunSink) conf.get(StormTopologyLayoutConstants.STREAMLINE_COMPONENT_CONF_KEY);
+        String spoutId = "testRunSinkBolt" + UUID_FOR_COMPONENTS;
```
Minor. Replace spout with bolt wherever applicable

HeartSaVioR replied:
Nice finding. Will fix.

priyank5485 commented on streams/runners/storm/layout/src/main/java/com/hortonworks/streamline/streams/layout/storm/TestRunSinkBoltFluxComponent.java:
```diff
@@ -0,0 +1,51 @@
+        } catch (JsonProcessingException e) {
+            log.error("Error creating json config string for TestRunSink", e);
+        }
+        constructorArgs.add(testRunSourceJson);
```
Why are we adding testRunSourceJson as constructor arg to test sink ?

priyank5485 replied:
Never mind. Ignore the above comment.

priyank5485 commented on streams/layout/src/main/java/com/hortonworks/streamline/streams/layout/component/impl/testing/TestRunSink.java:
```diff
@@ -0,0 +1,71 @@
+    }
+
+    public String getOutputFileUUID() {
+        return outputFileUUID;
```
Where is this being used? Should we remove it if its not used?

HeartSaVioR replied:
Yes we can remove this. Will address.

HeartSaVioR commented:
@priyank5485 Thanks for reviewing. Addressed review comments. Please let me know when you finished.

priyank5485 commented:
+1

HeartSaVioR commented:
@priyank5485 Addressed your comment on version id too. Also squashed.

HeartSaVioR commented:
Rebased.

HeartSaVioR commented:
There was a discussion from issue regarding mismatch between design & implementation and UI & UX (use case).
We agreed to change backend side, so some parts of design doc are now outdated, and so as API endpoints.

https://hwxiot.atlassian.net/browse/STREAMLINE-381?focusedCommentId=24808&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-24808

I changed the backend and rebased, but not yet updated description about API endpoints. I'll leave a comment regarding that.

HeartSaVioR commented:
API Endpoints
----

> POST /api/v1/catalog/topologies/:topologyId/testcases

>> Description

The endpoint adds test case for topology test run. Note that actual test records are stored to topology test run case source for each source.

>> Input parameters

* JSON String payload representing TopologyTestRunCase

```
{
  "name": "testcase1",
  "topologyId": 1
}
```

> PUT/GET/DELETE API endpoints for `testcases` available

> POST /api/v1/catalog/topologies/:topologyId/testcases/:testcaseId/sources

>> Description

The endpoint adds actual test records for specific test source on test case.

>> Input parameters

* JSON String payload representing TopologyTestRunCaseSource

```
{
  "testCaseId": 1,
  "sourceId": 1,
  "records": "{\"kafka_stream_1\": [{\"driverId\":24,\"truckId\":97,\"eventTime\":\"00:13.0\",\"eventType\":\"Normal\",\"longitude\":-88.77,\"latitude\":40.76,\"eventKey\":\"24|97|9223370572464762855\",\"correlationId\":\"3.66088E+18\",\"driverName\":\"Don Hilborn\",\"routeId\":1090292248,\"routeName\":\"Peoria to Ceder Rapids Route 2\",\"eventDate\":\"2016-05-27-22\",\"miles\":426},{\"driverId\":32,\"truckId\":42,\"eventTime\":\"00:13.1\",\"eventType\":\"Normal\",\"longitude\":-90.04,\"latitude\":35.19,\"eventKey\":\"32|42|9223370572464762755\",\"correlationId\":\"3.66088E+18\",\"driverName\":\"Ryan Templeton\",\"routeId\":1090292248,\"routeName\":\"Peoria to Ceder Rapids Route 2\",\"eventDate\":\"2016-05-27-22\",\"miles\":183},{\"driverId\":31,\"truckId\":18,\"eventTime\":\"00:13.1\",\"eventType\":\"Normal\",\"longitude\":-94.35,\"latitude\":38.33,\"eventKey\":\"31|18|9223370572464762726\",\"correlationId\":\"3.66088E+18\",\"driverName\":\"Rommel Garcia\",\"routeId\":1594289134,\"routeName\":\"Memphis to Little Rock Route 2\",\"eventDate\":\"2016-05-27-22\",\"miles\":34}]}"
}
```

> PUT/GET/LIST API endpoints for `testcases/sources` available (Do we need to have DELETE for sources?)

> GET /api/v1/catalog/topologies/:topologyId/testcases/:testcaseId/sources/topologysource/:topologySourceId

>> Similar as GET API endpoints for `testcases/sources`, but makes possible to get via topologySourceId rather than sourceId (topology test run case source's id)

> POST /api/v1/catalog/topologies/:topologyId/actions/testrun

>> Description

The endpoint runs given topology on test mode, and return the result of test run. Implementation details for running topology on test mode will be covered later.

>> Input parameters

* topologyId (Long)
* JSON String payload containing either test case ID, and expected output records (optional)

```
{
  “testCaseId”: 1,
  “expectedOutputRecords”: {
    "SINK1": [
      {"driverId": 20, "eventCount": 3, "maxLatitude": 39.78, ... },
      …
    ],
    ...
  },
}
```

"expectedOutputRecords" is optional. Only need to be provided when users want to compare actual output with expected output.
Assertion: All data sources should be covered. It will return HTTP status 400 if the assertion fails.

> GET /api/v1/catalog/topologies/:topologyId/(versions/:versionId)/actions/testrun/histories

>> Description

The endpoint retrieves the histories of topology test run for given topology, and return recent N history entities.

>> Input parameters

* topologyId (Long)
* limit (Int) - optional, will fix the default value

>> Output format (success case)

On succeed case, the response shows history entities matched to given criteria.

```
[
  {
    "id": 99,
    "topologyId": 1,
    "versionId": 1,
    "testRecords": "{\"1\":{\"kafka_stream_1\":[{\"driverId\":24,\"truckId\":97,\"eventTime\":\"00:13.0\",\"eventType\":\"Normal\",\"longitude\":-88.77,\"latitude\":40.76,\"eventKey\":\"24|97|9223370572464762855\",\"correlationId\":\"3.66088E+18\",\"driverName\":\"Don Hilborn\",\"routeId\":1090292248,\"routeName\":\"Peoria to Ceder Rapids Route 2\",\"eventDate\":\"2016-05-27-22\",\"miles\":426},{\"driverId\":32,\"truckId\":42,\"eventTime\":\"00:13.1\",\"eventType\":\"Normal\",\"longitude\":-90.04,\"latitude\":35.19,\"eventKey\":\"32|42|9223370572464762755\",\"correlationId\":\"3.66088E+18\",\"driverName\":\"Ryan Templeton\",\"routeId\":1090292248,\"routeName\":\"Peoria to Ceder Rapids Route 2\",\"eventDate\":\"2016-05-27-22\",\"miles\":183},{\"driverId\":31,\"truckId\":18,\"eventTime\":\"00:13.1\",\"eventType\":\"Normal\",\"longitude\":-94.35,\"latitude\":38.33,\"eventKey\":\"31|18|9223370572464762726\",\"correlationId\":\"3.66088E+18\",\"driverName\":\"Rommel Garcia\",\"routeId\":1594289134,\"routeName\":\"Memphis to Little Rock Route 2\",\"eventDate\":\"2016-05-27-22\",\"miles\":34}]},\"2\":{\"kafka_stream_2\":[{\"driverId\":24,\"truckId\":97,\"eventTime\":\"00:13.0\",\"eventType\":\"Normal\",\"longitude\":-88.77,\"latitude\":40.76,\"eventKey\":\"24|97|9223370572464762855\",\"correlationId\":\"3.66088E+18\",\"driverName\":\"Don Hilborn\",\"routeId\":1090292248,\"routeName\":\"Peoria to Ceder Rapids Route 2\",\"eventDate\":\"2016-05-27-22\",\"miles\":426},{\"driverId\":32,\"truckId\":42,\"eventTime\":\"00:13.1\",\"eventType\":\"Normal\",\"longitude\":-90.04,\"latitude\":35.19,\"eventKey\":\"32|42|9223370572464762755\",\"correlationId\":\"3.66088E+18\",\"driverName\":\"Ryan Templeton\",\"routeId\":1090292248,\"routeName\":\"Peoria to Ceder Rapids Route 2\",\"eventDate\":\"2016-05-27-22\",\"miles\":183},{\"driverId\":31,\"truckId\":18,\"eventTime\":\"00:13.1\",\"eventType\":\"Normal\",\"longitude\":-94.35,\"latitude\":38.33,\"eventKey\":\"31|18|9223370572464762726\",\"correlationId\":\"3.66088E+18\",\"driverName\":\"Rommel Garcia\",\"routeId\":1594289134,\"routeName\":\"Memphis to Little Rock Route 2\",\"eventDate\":\"2016-05-27-22\",\"miles\":34}]}}"
    "finish": true,
    "success": true,
    "expectedOutputRecords": "{\"KAFKA-SINK-1\":[{\"driverId\":24,\"truckId\":97,\"eventTime\":\"00:13.0\",\"eventType\":\"Normal\",\"longitude\":-88.77,\"latitude\":40.76,\"eventKey\":\"24|97|9223370572464762855\",\"correlationId\":\"3.66088E+18\",\"driverName\":\"Don Hilborn\",\"routeId\":1090292248,\"routeName\":\"Peoria to Ceder Rapids Route 2\",\"eventDate\":\"2016-05-27-22\",\"miles\":426},{\"driverId\":32,\"truckId\":42,\"eventTime\":\"00:13.1\",\"eventType\":\"Normal\",\"longitude\":-90.04,\"latitude\":35.19,\"eventKey\":\"32|42|9223370572464762755\",\"correlationId\":\"3.66088E+18\",\"driverName\":\"Ryan Templeton\",\"routeId\":1090292248,\"routeName\":\"Peoria to Ceder Rapids Route 2\",\"eventDate\":\"2016-05-27-22\",\"miles\":183},{\"driverId\":31,\"truckId\":18,\"eventTime\":\"00:13.1\",\"eventType\":\"Normal\",\"longitude\":-94.35,\"latitude\":38.33,\"eventKey\":\"31|18|9223370572464762726\",\"correlationId\":\"3.66088E+18\",\"driverName\":\"Rommel Garcia\",\"routeId\":1594289134,\"routeName\":\"Memphis to Little Rock Route 2\",\"eventDate\":\"2016-05-27-22\",\"miles\":34}],\"KAFKA-1\":[{\"driverId\":24,\"truckId\":97,\"eventTime\":\"00:13.0\",\"eventType\":\"Normal\",\"longitude\":-88.77,\"latitude\":40.76,\"eventKey\":\"24|97|9223370572464762855\",\"correlationId\":\"3.66088E+18\",\"driverName\":\"Don Hilborn\",\"routeId\":1090292248,\"routeName\":\"Peoria to Ceder Rapids Route 2\",\"eventDate\":\"2016-05-27-22\",\"miles\":426},{\"driverId\":32,\"truckId\":42,\"eventTime\":\"00:13.1\",\"eventType\":\"Normal\",\"longitude\":-90.04,\"latitude\":35.19,\"eventKey\":\"32|42|9223370572464762755\",\"correlationId\":\"3.66088E+18\",\"driverName\":\"Ryan Templeton\",\"routeId\":1090292248,\"routeName\":\"Peoria to Ceder Rapids Route 2\",\"eventDate\":\"2016-05-27-22\",\"miles\":183},{\"driverId\":31,\"truckId\":18,\"eventTime\":\"00:13.1\",\"eventType\":\"Normal\",\"longitude\":-94.35,\"latitude\":38.33,\"eventKey\":\"31|18|9223370572464762726\",\"correlationId\":\"3.66088E+18\",\"driverName\":\"Rommel Garcia\",\"routeId\":1594289134,\"routeName\":\"Memphis to Little Rock Route 2\",\"eventDate\":\"2016-05-27-22\",\"miles\":34}]}",
    "actualOutputRecords": "{\"KAFKA-SINK-1\":[{\"driverId\":24,\"truckId\":97,\"eventTime\":\"00:13.0\",\"eventType\":\"Normal\",\"longitude\":-88.77,\"latitude\":40.76,\"eventKey\":\"24|97|9223370572464762855\",\"correlationId\":\"3.66088E+18\",\"driverName\":\"Don Hilborn\",\"routeId\":1090292248,\"routeName\":\"Peoria to Ceder Rapids Route 2\",\"eventDate\":\"2016-05-27-22\",\"miles\":426},{\"driverId\":32,\"truckId\":42,\"eventTime\":\"00:13.1\",\"eventType\":\"Normal\",\"longitude\":-90.04,\"latitude\":35.19,\"eventKey\":\"32|42|9223370572464762755\",\"correlationId\":\"3.66088E+18\",\"driverName\":\"Ryan Templeton\",\"routeId\":1090292248,\"routeName\":\"Peoria to Ceder Rapids Route 2\",\"eventDate\":\"2016-05-27-22\",\"miles\":183},{\"driverId\":31,\"truckId\":18,\"eventTime\":\"00:13.1\",\"eventType\":\"Normal\",\"longitude\":-94.35,\"latitude\":38.33,\"eventKey\":\"31|18|9223370572464762726\",\"correlationId\":\"3.66088E+18\",\"driverName\":\"Rommel Garcia\",\"routeId\":1594289134,\"routeName\":\"Memphis to Little Rock Route 2\",\"eventDate\":\"2016-05-27-22\",\"miles\":34}],\"KAFKA-1\":[{\"driverId\":24,\"truckId\":97,\"eventTime\":\"00:13.0\",\"eventType\":\"Normal\",\"longitude\":-88.77,\"latitude\":40.76,\"eventKey\":\"24|97|9223370572464762855\",\"correlationId\":\"3.66088E+18\",\"driverName\":\"Don Hilborn\",\"routeId\":1090292248,\"routeName\":\"Peoria to Ceder Rapids Route 2\",\"eventDate\":\"2016-05-27-22\",\"miles\":426},{\"driverId\":32,\"truckId\":42,\"eventTime\":\"00:13.1\",\"eventType\":\"Normal\",\"longitude\":-90.04,\"latitude\":35.19,\"eventKey\":\"32|42|9223370572464762755\",\"correlationId\":\"3.66088E+18\",\"driverName\":\"Ryan Templeton\",\"routeId\":1090292248,\"routeName\":\"Peoria to Ceder Rapids Route 2\",\"eventDate\":\"2016-05-27-22\",\"miles\":183},{\"driverId\":31,\"truckId\":18,\"eventTime\":\"00:13.1\",\"eventType\":\"Normal\",\"longitude\":-94.35,\"latitude\":38.33,\"eventKey\":\"31|18|9223370572464762726\",\"correlationId\":\"3.66088E+18\",\"driverName\":\"Rommel Garcia\",\"routeId\":1594289134,\"routeName\":\"Memphis to Little Rock Route 2\",\"eventDate\":\"2016-05-27-22\",\"miles\":34}]}",
    “matched” : true,
    "startTime": 1234567890,
    "finishTime": 2345678901,
    "timestamp": 2345678901
  },
  …
]
```

HeartSaVioR commented:
@shahsank3t 
Please have a look at updated API endpoint.
Please note that I don't address having simple output version of topology test run history API endpoint, since topology test run is running synchronously. If we really want to make it asynchronous, then I'll address, but maybe better to address from a new issue.

@satishd @priyank5485 
Sorry for rebasing earlier. Please review updated PR again if you have time. Thanks in advance!

HeartSaVioR commented:
@shahsank3t 
Updated PR to reflect your request on allowing individual storing of expected records on each sink.
Below is updated API Endpoints.

API Endpoints
----

> POST /api/v1/catalog/topologies/:topologyId/testcases

>> Description

The endpoint adds test case for topology test run. Note that actual test records are stored to topology test run case source for each source.

>> Input parameters

* JSON String payload representing TopologyTestRunCase

```
{
  "name": "testcase1",
  "topologyId": 1
}
```

> PUT/GET/DELETE API endpoints for `testcases` available

> POST /api/v1/catalog/topologies/:topologyId/testcases/:testcaseId/sources

>> Description

The endpoint adds actual test records for specific test source on test case.

>> Input parameters

* JSON String payload representing TopologyTestRunCaseSource

```
{
  "testCaseId": 1,
  "sourceId": 1,
  "records": "{\"kafka_stream_1\": [{\"driverId\":24,\"truckId\":97,\"eventTime\":\"00:13.0\",\"eventType\":\"Normal\",\"longitude\":-88.77,\"latitude\":40.76,\"eventKey\":\"24|97|9223370572464762855\",\"correlationId\":\"3.66088E+18\",\"driverName\":\"Don Hilborn\",\"routeId\":1090292248,\"routeName\":\"Peoria to Ceder Rapids Route 2\",\"eventDate\":\"2016-05-27-22\",\"miles\":426},{\"driverId\":32,\"truckId\":42,\"eventTime\":\"00:13.1\",\"eventType\":\"Normal\",\"longitude\":-90.04,\"latitude\":35.19,\"eventKey\":\"32|42|9223370572464762755\",\"correlationId\":\"3.66088E+18\",\"driverName\":\"Ryan Templeton\",\"routeId\":1090292248,\"routeName\":\"Peoria to Ceder Rapids Route 2\",\"eventDate\":\"2016-05-27-22\",\"miles\":183},{\"driverId\":31,\"truckId\":18,\"eventTime\":\"00:13.1\",\"eventType\":\"Normal\",\"longitude\":-94.35,\"latitude\":38.33,\"eventKey\":\"31|18|9223370572464762726\",\"correlationId\":\"3.66088E+18\",\"driverName\":\"Rommel Garcia\",\"routeId\":1594289134,\"routeName\":\"Memphis to Little Rock Route 2\",\"eventDate\":\"2016-05-27-22\",\"miles\":34}]}"
}
```

> PUT/GET/LIST API endpoints for `testcases/sources` available (Do we need to have DELETE for sources?)

> GET /api/v1/catalog/topologies/:topologyId/testcases/:testcaseId/sources/topologysource/:topologySourceId

>> Similar as GET API endpoints for `testcases/sources`, but makes possible to get via topologySourceId rather than sourceId (topology test run case source's id)


> POST /api/v1/catalog/topologies/:topologyId/testcases/:testcaseId/sinks

>> Description

The endpoint adds expected records for specific test sink on test case.

>> Input parameters

* JSON String payload representing TopologyTestRunCaseSink

```
{
  "testCaseId": 2,
  "sinkId": 10,
  "records": "[{\"driverId\":24,\"truckId\":97,\"eventTime\":\"00:13.0\",\"eventType\":\"Normal\",\"longitude\":-88.77,\"latitude\":40.76,\"eventKey\":\"24|97|9223370572464762855\",\"correlationId\":\"3.66088E+18\",\"driverName\":\"Don Hilborn\",\"routeId\":1090292248,\"routeName\":\"Peoria to Ceder Rapids Route 2\",\"eventDate\":\"2016-05-27-22\",\"miles\":426},{\"driverId\":32,\"truckId\":42,\"eventTime\":\"00:13.1\",\"eventType\":\"Normal\",\"longitude\":-90.04,\"latitude\":35.19,\"eventKey\":\"32|42|9223370572464762755\",\"correlationId\":\"3.66088E+18\",\"driverName\":\"Ryan Templeton\",\"routeId\":1090292248,\"routeName\":\"Peoria to Ceder Rapids Route 2\",\"eventDate\":\"2016-05-27-22\",\"miles\":183},{\"driverId\":31,\"truckId\":18,\"eventTime\":\"00:13.1\",\"eventType\":\"Normal\",\"longitude\":-94.35,\"latitude\":38.33,\"eventKey\":\"31|18|9223370572464762726\",\"correlationId\":\"3.66088E+18\",\"driverName\":\"Rommel Garcia\",\"routeId\":1594289134,\"routeName\":\"Memphis to Little Rock Route 2\",\"eventDate\":\"2016-05-27-22\",\"miles\":34}]"
}
```

> PUT/GET/LIST API endpoints for `testcases/sinks` available (Do we need to have DELETE for sinks?)

> GET /api/v1/catalog/topologies/:topologyId/testcases/:testcaseId/sinks/topologysink/:topologySinkId

>> Similar as GET API endpoints for `testcases/sinks`, but makes possible to get via topologySinkId rather than sinkId (topology test run case sink's id)

> POST /api/v1/catalog/topologies/:topologyId/actions/testrun

>> Description

The endpoint runs given topology on test mode, and return the result of test run. Implementation details for running topology on test mode will be covered later.

>> Input parameters

* topologyId (Long)
* JSON String payload containing either test case ID

```
{
  “testCaseId”: 1
}
```

Assertion: All data sources should be covered. It will return HTTP status 400 if the assertion fails.

> GET /api/v1/catalog/topologies/:topologyId/(versions/:versionId)/actions/testrun/histories

>> Description

The endpoint retrieves the histories of topology test run for given topology, and return recent N history entities.

>> Input parameters

* topologyId (Long)
* limit (Int) - optional, will fix the default value

>> Output format (success case)

On succeed case, the response shows history entities matched to given criteria.

```
[
  {
    "id": 99,
    "topologyId": 1,
    "versionId": 1,
    "testRecords": "{\"1\":{\"kafka_stream_1\":[{\"driverId\":24,\"truckId\":97,\"eventTime\":\"00:13.0\",\"eventType\":\"Normal\",\"longitude\":-88.77,\"latitude\":40.76,\"eventKey\":\"24|97|9223370572464762855\",\"correlationId\":\"3.66088E+18\",\"driverName\":\"Don Hilborn\",\"routeId\":1090292248,\"routeName\":\"Peoria to Ceder Rapids Route 2\",\"eventDate\":\"2016-05-27-22\",\"miles\":426},{\"driverId\":32,\"truckId\":42,\"eventTime\":\"00:13.1\",\"eventType\":\"Normal\",\"longitude\":-90.04,\"latitude\":35.19,\"eventKey\":\"32|42|9223370572464762755\",\"correlationId\":\"3.66088E+18\",\"driverName\":\"Ryan Templeton\",\"routeId\":1090292248,\"routeName\":\"Peoria to Ceder Rapids Route 2\",\"eventDate\":\"2016-05-27-22\",\"miles\":183},{\"driverId\":31,\"truckId\":18,\"eventTime\":\"00:13.1\",\"eventType\":\"Normal\",\"longitude\":-94.35,\"latitude\":38.33,\"eventKey\":\"31|18|9223370572464762726\",\"correlationId\":\"3.66088E+18\",\"driverName\":\"Rommel Garcia\",\"routeId\":1594289134,\"routeName\":\"Memphis to Little Rock Route 2\",\"eventDate\":\"2016-05-27-22\",\"miles\":34}]},\"2\":{\"kafka_stream_2\":[{\"driverId\":24,\"truckId\":97,\"eventTime\":\"00:13.0\",\"eventType\":\"Normal\",\"longitude\":-88.77,\"latitude\":40.76,\"eventKey\":\"24|97|9223370572464762855\",\"correlationId\":\"3.66088E+18\",\"driverName\":\"Don Hilborn\",\"routeId\":1090292248,\"routeName\":\"Peoria to Ceder Rapids Route 2\",\"eventDate\":\"2016-05-27-22\",\"miles\":426},{\"driverId\":32,\"truckId\":42,\"eventTime\":\"00:13.1\",\"eventType\":\"Normal\",\"longitude\":-90.04,\"latitude\":35.19,\"eventKey\":\"32|42|9223370572464762755\",\"correlationId\":\"3.66088E+18\",\"driverName\":\"Ryan Templeton\",\"routeId\":1090292248,\"routeName\":\"Peoria to Ceder Rapids Route 2\",\"eventDate\":\"2016-05-27-22\",\"miles\":183},{\"driverId\":31,\"truckId\":18,\"eventTime\":\"00:13.1\",\"eventType\":\"Normal\",\"longitude\":-94.35,\"latitude\":38.33,\"eventKey\":\"31|18|9223370572464762726\",\"correlationId\":\"3.66088E+18\",\"driverName\":\"Rommel Garcia\",\"routeId\":1594289134,\"routeName\":\"Memphis to Little Rock Route 2\",\"eventDate\":\"2016-05-27-22\",\"miles\":34}]}}"
    "finish": true,
    "success": true,
    "expectedOutputRecords": "{\"KAFKA-SINK-1\":[{\"driverId\":24,\"truckId\":97,\"eventTime\":\"00:13.0\",\"eventType\":\"Normal\",\"longitude\":-88.77,\"latitude\":40.76,\"eventKey\":\"24|97|9223370572464762855\",\"correlationId\":\"3.66088E+18\",\"driverName\":\"Don Hilborn\",\"routeId\":1090292248,\"routeName\":\"Peoria to Ceder Rapids Route 2\",\"eventDate\":\"2016-05-27-22\",\"miles\":426},{\"driverId\":32,\"truckId\":42,\"eventTime\":\"00:13.1\",\"eventType\":\"Normal\",\"longitude\":-90.04,\"latitude\":35.19,\"eventKey\":\"32|42|9223370572464762755\",\"correlationId\":\"3.66088E+18\",\"driverName\":\"Ryan Templeton\",\"routeId\":1090292248,\"routeName\":\"Peoria to Ceder Rapids Route 2\",\"eventDate\":\"2016-05-27-22\",\"miles\":183},{\"driverId\":31,\"truckId\":18,\"eventTime\":\"00:13.1\",\"eventType\":\"Normal\",\"longitude\":-94.35,\"latitude\":38.33,\"eventKey\":\"31|18|9223370572464762726\",\"correlationId\":\"3.66088E+18\",\"driverName\":\"Rommel Garcia\",\"routeId\":1594289134,\"routeName\":\"Memphis to Little Rock Route 2\",\"eventDate\":\"2016-05-27-22\",\"miles\":34}],\"KAFKA-1\":[{\"driverId\":24,\"truckId\":97,\"eventTime\":\"00:13.0\",\"eventType\":\"Normal\",\"longitude\":-88.77,\"latitude\":40.76,\"eventKey\":\"24|97|9223370572464762855\",\"correlationId\":\"3.66088E+18\",\"driverName\":\"Don Hilborn\",\"routeId\":1090292248,\"routeName\":\"Peoria to Ceder Rapids Route 2\",\"eventDate\":\"2016-05-27-22\",\"miles\":426},{\"driverId\":32,\"truckId\":42,\"eventTime\":\"00:13.1\",\"eventType\":\"Normal\",\"longitude\":-90.04,\"latitude\":35.19,\"eventKey\":\"32|42|9223370572464762755\",\"correlationId\":\"3.66088E+18\",\"driverName\":\"Ryan Templeton\",\"routeId\":1090292248,\"routeName\":\"Peoria to Ceder Rapids Route 2\",\"eventDate\":\"2016-05-27-22\",\"miles\":183},{\"driverId\":31,\"truckId\":18,\"eventTime\":\"00:13.1\",\"eventType\":\"Normal\",\"longitude\":-94.35,\"latitude\":38.33,\"eventKey\":\"31|18|9223370572464762726\",\"correlationId\":\"3.66088E+18\",\"driverName\":\"Rommel Garcia\",\"routeId\":1594289134,\"routeName\":\"Memphis to Little Rock Route 2\",\"eventDate\":\"2016-05-27-22\",\"miles\":34}]}",
    "actualOutputRecords": "{\"KAFKA-SINK-1\":[{\"driverId\":24,\"truckId\":97,\"eventTime\":\"00:13.0\",\"eventType\":\"Normal\",\"longitude\":-88.77,\"latitude\":40.76,\"eventKey\":\"24|97|9223370572464762855\",\"correlationId\":\"3.66088E+18\",\"driverName\":\"Don Hilborn\",\"routeId\":1090292248,\"routeName\":\"Peoria to Ceder Rapids Route 2\",\"eventDate\":\"2016-05-27-22\",\"miles\":426},{\"driverId\":32,\"truckId\":42,\"eventTime\":\"00:13.1\",\"eventType\":\"Normal\",\"longitude\":-90.04,\"latitude\":35.19,\"eventKey\":\"32|42|9223370572464762755\",\"correlationId\":\"3.66088E+18\",\"driverName\":\"Ryan Templeton\",\"routeId\":1090292248,\"routeName\":\"Peoria to Ceder Rapids Route 2\",\"eventDate\":\"2016-05-27-22\",\"miles\":183},{\"driverId\":31,\"truckId\":18,\"eventTime\":\"00:13.1\",\"eventType\":\"Normal\",\"longitude\":-94.35,\"latitude\":38.33,\"eventKey\":\"31|18|9223370572464762726\",\"correlationId\":\"3.66088E+18\",\"driverName\":\"Rommel Garcia\",\"routeId\":1594289134,\"routeName\":\"Memphis to Little Rock Route 2\",\"eventDate\":\"2016-05-27-22\",\"miles\":34}],\"KAFKA-1\":[{\"driverId\":24,\"truckId\":97,\"eventTime\":\"00:13.0\",\"eventType\":\"Normal\",\"longitude\":-88.77,\"latitude\":40.76,\"eventKey\":\"24|97|9223370572464762855\",\"correlationId\":\"3.66088E+18\",\"driverName\":\"Don Hilborn\",\"routeId\":1090292248,\"routeName\":\"Peoria to Ceder Rapids Route 2\",\"eventDate\":\"2016-05-27-22\",\"miles\":426},{\"driverId\":32,\"truckId\":42,\"eventTime\":\"00:13.1\",\"eventType\":\"Normal\",\"longitude\":-90.04,\"latitude\":35.19,\"eventKey\":\"32|42|9223370572464762755\",\"correlationId\":\"3.66088E+18\",\"driverName\":\"Ryan Templeton\",\"routeId\":1090292248,\"routeName\":\"Peoria to Ceder Rapids Route 2\",\"eventDate\":\"2016-05-27-22\",\"miles\":183},{\"driverId\":31,\"truckId\":18,\"eventTime\":\"00:13.1\",\"eventType\":\"Normal\",\"longitude\":-94.35,\"latitude\":38.33,\"eventKey\":\"31|18|9223370572464762726\",\"correlationId\":\"3.66088E+18\",\"driverName\":\"Rommel Garcia\",\"routeId\":1594289134,\"routeName\":\"Memphis to Little Rock Route 2\",\"eventDate\":\"2016-05-27-22\",\"miles\":34}]}",
    “matched” : true,
    "startTime": 1234567890,
    "finishTime": 2345678901,
    "timestamp": 2345678901
  },
  …
]
```

HeartSaVioR commented:
#536 is follow-up version of this PR. I don't close this right now, but if we're OK to move on #536 I'll close this.
